clang/runtime/dpct-rt/include/kernel.hpp.inc

```diff
@@ -86,5 +111,239 @@ static kernel_function_info get_kernel_function_info(const void *function) {
+      pe_header + coff_header_size + sizeof_optional_header;
+  if (extract64(tail_of_optional_header - 8) != 0) {
+    throw std::runtime_error("Optional header not zero-padded");
+  }
+
+  ///////////////////////////////////////////////////////////////////////
+  // Analyze last section header
+  constexpr int section_header_size = 40;
+  unsigned char const *const last_section_header =
+      tail_of_optional_header + section_header_size * (number_of_sections - 1);
+
+  uint32_t sizeof_raw_data = extract32(last_section_header + 16);
+  uint32_t pointer_to_raw_data = extract32(last_section_header + 20);
+
+  return sizeof_raw_data + pointer_to_raw_data;
+#else
+  if (blob[0] != 0x7F || blob[1] != 'E' || blob[2] != 'L' || blob[3] != 'F')
+    throw std::runtime_error("Blob is not in ELF format");
+
+  if (blob[4] != 0x02)
+    throw std::runtime_error("Only 64-bit headers are supported");
+
+  if (blob[5] != 0x01)
+    throw std::runtime_error("Only little-endian headers are supported");
+
+  unsigned char const *const ublob =
+      reinterpret_cast<unsigned char const *const>(blob);
+  uint64_t e_shoff = extract64(ublob + 0x28);
+  uint16_t e_shentsize = extract16(ublob + 0x3A);
+  uint16_t e_shnum = extract16(ublob + 0x3C);
+
+  return e_shoff + (e_shentsize * e_shnum);
+#endif
+}
+
+} // namespace detail
+
+class module {
+public:
+  module() : ptr{nullptr} {}
+  module(void *ptr) : ptr{ptr} {}
+
+  operator void *() const { return ptr; }
+
+  void *get_ptr() { return (ptr); }
+
+private:
+  void *ptr;
+};
+
+namespace detail {
+
+static std::string mkfilename() {
+  time_t rawtime;
+  time(&rawtime);
+
+  std::stringstream buffer;
+  buffer << static_cast<unsigned long long>(rawtime);
+
+  std::hash<std::string> hasher;
+  int i = hasher(buffer.str());
+  static int cnt;
+
+  std::stringstream filename;
+
+  filename << std::filesystem::temp_directory_path().string();
+#ifndef _WIN32
+  filename << "/dpct";
+#else
+  filename << "dpct";
+#endif
+  filename << std::setfill('0') << std::setw(2) << std::right << std::hex
+           << ++cnt;
+  filename << std::setfill('0') << std::setw(8) << std::right << std::hex << i;
+#ifdef _WIN32
+  filename << ".dll";
+#endif
+
+  return (filename.str());
+}
+
+static module load_dl_from_data(char const *const data, size_t size) {
+  const std::string name = mkfilename();
+
+  write_data_to_file(name, data, size);
+#ifdef _WIN32
+  void *so = LoadLibraryA(name.c_str());
+#else
+  void *so = dlopen(name.c_str(), RTLD_LAZY);
+#endif
+  if (so == nullptr)
+    throw std::runtime_error("Failed to load module");
+
+  std::remove(name.c_str());
+
+  return so;
+}
+
+} // namespace detail
+
+static module load_sycl_lib(const std::string &name) {
+  std::ifstream ifs;
+  ifs.open(name, std::ios::in | std::ios::binary);
+
+  std::stringstream buffer;
+  buffer << ifs.rdbuf();
+  return detail::load_dl_from_data(buffer.str().c_str(), buffer.str().size());
+};
```

Comment:
`buffer.str()` returns a copy of the `buffer`'s underlying string, so the buffer is copied an extra time during `buffer.str().size()`. Also, since it's copied, care for dangling pointer from `.c_str()` (fine now since the temporary `buffer.str()` is destructed after `load_dl_from_data`).

Reply:
Fixed.